runtime/builtin_types.go

```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return Abs(f, args[0])
 }
 
+func zipLongestImpl(f *Frame, args Args) ([]*Object, *BaseException) {
+	argc := len(args)
+	result := make([]*Object, 0, 2)
+	iters := make([]*Object, argc)
+	for i, arg := range args {
+		iter, raised := Iter(f, arg)
+		if raised != nil {
+			return nil, raised
+		}
+		iters[i] = iter
+	}
+
+	for {
+		noItems := true
+		elems := make([]*Object, argc)
+		for i, iter := range iters {
+			elem, raised := Next(f, iter)
+			if raised != nil {
+				if raised.isInstance(StopIterationType) {
+					continue
+				}
+				return nil, raised
+			}
+			noItems = false
+			elems[i] = elem
+		}
+		if noItems {
+			break
+		}
+		// Replace nil with None
+		if len(elems) > 1 {
+			lastIndex := len(elems) - 1
+			for elems[lastIndex] == nil {
+				elems[lastIndex] = None
+				lastIndex--
+			}
+		}
+		// NOTE: append does the capacity doubling logic as needed
+		result = append(result, NewTuple(elems...).ToObject())
+	}
+	return result, nil
+}
+
+func builtinMapFn(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc < 2 {
+		return nil, f.RaiseType(TypeErrorType, "map() requires at least two args")
+	}
+
+	pred := toFunctionUnsafe(args[0])
+	result := make([]*Object, 0, 2)
+
+	z, raised := zipLongestImpl(f, args[1:])
+	if raised != nil {
+		return nil, raised
+	}
+	for _, tuple := range z {
+		if pred.Type() == NoneType {
+			if argc == 2 {
+				result = append(result, toTupleUnsafe(tuple).GetItem(0))
+			} else {
+				result = append(result, tuple)
+			}
+		} else {
+			ret, raised := pred.Call(f, toTupleUnsafe(tuple).elems, nil)
```

Comment:
Rather than allocating a Tuple and then not using it in most cases, maybe zipLongest should return `[][]*Object.`